### DISCUSSION TITLE
How to determine object size?

### DISCUSSION THREAD

**Discussion by gnbl** [micropython/micropython#9945] [Using MicroPython] [answered]
Does MicroPython have a method to determine object size like `sys.getsizeof(_object_)`?
I'd like to compare different data structures for memory efficiency (look up a couple thousand timetable entries, for example search in a tuple vs. dictionary).

**Comment by jimmo** [comment, marked as answer]
`sys.getsizeof` is not implemented in MicroPython. Internally, the heap does know the allocated size of a given object, but many objects have multiple layers of allocations (e.g. a dictionary object contains a pointer to its table, which contains pointers to keys and values, etc).

`getsizeof` works in CPython with runtime support - builtin types must implement this to report their size. This would add a lot of code size, so is unlikely to ever be included in standard MicroPython firmware.

If your goal is to understand how much RAM different representations take up, then using `gc.mem_alloc()` to compute a difference might be more useful.

```py
def make_timetable():
  ...
  return obj

def compute_alloc_diff():
  gc.collect()
  before = gc.mem_alloc()
  x = make_timetable()
  gc.collect()
  print(gc.mem_alloc() - before)
```

Note that `make_timetable` is in a function so that any local variables go out of scope -- only the final object will be still "live" when gc.collect() is called the second time.

Some caveats though:
 - Ensure that `make_timetable` doesn't mutate any global state. e.g. modifying global objects.
 - It's a bit hard to account for string data, due to string interning, which are stored in chunks. Interned strings are never freed though, so one way to account for this is to call `compute_alloc_diff` twice. The second time will not include any string data, and only the actual data structures.

**Reply by gnbl** [reply]
Thanks again.
It looks like tuples might be read from (and kept in) flash memory since `compute_alloc_diff()` returns zero.
But a binary search (index of next larger element) seems to work nicely on that data (ordered timestamps), so I'll probably not investigate further (const dict or multi-dimensional list).

**Reply by jimmo** [reply]
Yes, if you're using frozen code, then tuples are constant objects and stored in ROM/flash.
